once/_iterator_wrappers.py

```diff
@@ -0,0 +1,172 @@
+    YIELDING = 2
+    # Waiting for the underlying iterator, already triggered from another call.
+    WAITING = 3
+
+
+class GeneratorWrapper:
+    """Wrapper around an sync generator which only runs once.
+
+    Subsequent calls will return results from the first call, which is
+    evaluated lazily.
+    """
+
+    def __init__(self, func, *args, **kwargs) -> None:
+        self.generator: collections.abc.Generator | None = func(*args, **kwargs)
+        self.finished = False
+        self.results: list = []
+        self.generating = False
+        self.lock = threading.Lock()
+        self.next_send = None
+
+    def yield_results(self) -> collections.abc.Generator:
+        # Fast path for subsequent repeated call:
+        with self.lock:
+            finished = self.finished
+        if finished:
+            yield from self.results
+            return
+        i = 0
+        yield_value = None
+        next_send = None
+        # Fast path for subsequent calls will not require a lock
+        while True:
+            action: _IteratorAction | None = None
+            # With a lock, we figure out which action to take, and then we take it after release.
+            with self.lock:
+                if i == len(self.results):
+                    if self.finished:
+                        return
+                    if self.generating:
+                        action = _IteratorAction.WAITING
+                    else:
+                        action = _IteratorAction.GENERATING
+                        next_send = self.next_send
+                        self.generating = True
+                else:
+                    action = _IteratorAction.YIELDING
+                    yield_value = self.results[i]
+            if action == _IteratorAction.WAITING:
+                continue
+            if action == _IteratorAction.YIELDING:
+                next_send = yield yield_value
+                i += 1
+                # If we have just sent the last element and we have not yet kicked off the next
+                # iteration, we need to record the next send value.
+                with self.lock:
+                    if i == len(self.results) and not self.generating:
+                        self.next_send = next_send
+                continue
+            if action == _IteratorAction.GENERATING:
+                assert self.generator is not None
+                try:
+                    result = self.generator.send(next_send)
+                except StopIteration:
+                    # This lock should be unnecessary, which by definition means there should be no
+                    # contention on it, so we use it to preserve our assumptions about variables which
+                    # are modified under lock.
+                    with self.lock:
+                        self.finished = True
+                        self.generator = None  # Allow this to be GCed.
+                        self.generating = False
+                else:
```

Comment:
nit: It's hard to tell if this `else` is on the `if` or the `try`. I think you could just in-line the else below `result = self.generator.send(next_send)`, like:

```python
try:
    result = self.generator.send(next_send)
    with self.lock:
        self.results.append(result)
        self.generating = False
except StopIteration:
    ...
```

Reply:
is this any clearer?